newlangref: add symbols to spec (#385)
numba/numba-mlir: doc/newlangref.md

            d += tmp * tmp
        D[i, j] = np.sqrt(d)

# New api, immediately switching to workitem level.
W1 = sym.W1
W2 = sym.W2
H = sym.H

> @kernel(work_size=(W1, W2))

Reviewer: ```suggestion
@kernel(work_shape=(W1, W2))
```

    inner()

# Using WG level api

> @kernel(work_size=(W1, W2))

Reviewer: ```suggestion
@kernel(work_shape=(W1, W2))
```

# Current kernel API
pairwise_distance_kernel[global_size, local_size](X1, X2, D)

> # In New API, work, group and subgroup sizes are bound to kernel params via

Reviewer: ```suggestion
# In New API work/group shapes and subgroup size are bound to kernel params via
```

> # symbols, i.e. in previous example they are directly taken from input buffers

Reviewer: ```suggestion
# symbols, i.e. in previous example they are taken from input buffers
```

# dimenstions.
pairwise_distance_kernel(X1, X2, D)
```
While kernel function takes `CurrentGroup` as argument, it's not passed to the
kernel invocation directly and group dimension are iferred from bound symbols.

> If user need to pass dimensions unrelated to any buffer args, he can bind them

Reviewer: ```suggestion
If user wants to specify work/group sizes explicitly they may bind it
```

> to the separate kernel int/tuple arg.

Reviewer: ```suggestion
to (tuples of) symbols passed as kernel arguments
```

```python
@kernel(work_shape=(G1,G2,G3), group_shape=(L1,L2,L3))
def test(gr: CurrentGroup,
         gsize: tuple[G1, G2, G3],
         lsize: tuple[L1, L2, L3]):
    # gsize and lsize are not used inside kernel and only needed to bind
    # work and group shape.
    ...

test((1024, 1, 1), (64, 1, 1))
```

### Symbols
Symbols are the way to define some relations between input buffers dimensions
and/or work or group shape:
```python
W, H = sym.W, sym.H

> @kernel(work_size=(W, H))

Reviewer: ```suggestion
@kernel(work_shape=(W, H))
```

def pairwise_distance_kernel(group: CurrentGroup,
                             X1: Buffer[W, H],
                             X2: Buffer[W, H]):
    # Kernel expects 2 buffers with same shape and work size will be equal to
    # that shape
    ...
```

By default symbols are treated as dynamic values, i.e. in previous example there
will be a single kernel for every input arrays size.

Symbols can also be declared as literals, and in this case, runtime will compile
separate versions of the kernel for each distinct symbol value:
```python

> # H is usually small and won't change between kernel invocations, declare is as

Reviewer: ```suggestion
# H is usually small and won't change between kernel invocations, declaring it as
```
